Cy8cproto_041tp SPI driver implementation

dharunkrithikk opened the pull request:
This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/100894

dharunkrithikk committed:
boards: cy8cproto_041tp: enable SPI1

Enable SCB1 as SPI master and configure the corresponding peripheral
clock resource.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk committed:
dts: infineon: add PSoC4100TP SPI pinctrl definitions

Add pin control definitions for SPI master signals (CLK, MISO, MOSI,
and SELECT lines) for SCB0 and SCB1 on the PSoC4100TP 64-TQFP package.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sreeramIfx commented on drivers/spi/spi_infineon_pdl.c:
```diff
@@ -25,13 +25,10 @@ LOG_MODULE_REGISTER(cat1_spi, CONFIG_SPI_LOG_LEVEL);
-#define IFX_CAT1_SPI_MAX_DATA_WIDTH (32)
+#define IFX_CAT1_SPI_DEFAULT_OVERSAMPLE CONFIG_IFX_CAT1_SPI_OVERSAMPLE_DEFAULT
+
+#define IFX_CAT1_SPI_MIN_DATA_WIDTH CONFIG_IFX_CAT1_SPI_MIN_DATA_WIDTH
```
Can we use CONFIG_IFX_CAT1_SPI_MIN_DATA_WIDTH directly?

dharunkrithikk replied:
changed

sreeramIfx commented on drivers/spi/spi_infineon_pdl.c:
```diff
@@ -570,6 +716,20 @@ static int ifx_cat1_spi_init(const struct device *dev)
+ * (matching bare-metal implementation)
+ */
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
+#define DEFAULT_TX_FIFO_TRIGGER_LEVEL 7
```
why is this specific to PSOC4?

dharunkrithikk committed:
samples: bme280: add overlay for cy8cproto_041tp

Add board overlay for Infineon CY8CPROTO-041TP to support the
BME280 sensor sample. The configuration enables SPI1 and connects
the BME280 sensor with Chip Select on P4.0.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk commented on drivers/spi/spi_infineon_pdl.c:
```diff
@@ -50,6 +47,12 @@ LOG_MODULE_REGISTER(cat1_spi, CONFIG_SPI_LOG_LEVEL);
 #endif
 
+#if defined(CY_IP_MXSCB_INSTANCES)
+#define _IFX_SCB_ARRAY_SIZE (CY_IP_MXSCB_INSTANCES)
```
changed

dharunkrithikk commented on drivers/spi/spi_infineon_pdl.c:
```diff
@@ -50,6 +47,12 @@ LOG_MODULE_REGISTER(cat1_spi, CONFIG_SPI_LOG_LEVEL);
+#if defined(CY_IP_MXSCB_INSTANCES)
+#define _IFX_SCB_ARRAY_SIZE (CY_IP_MXSCB_INSTANCES)
+#elif defined(CY_IP_M0S8SCB_INSTANCES)
+#define _IFX_SCB_ARRAY_SIZE (CY_IP_M0S8SCB_INSTANCES)
```
changed

sreeramIfx commented:
Please fix the order of the commits. Overlays are introduced before the driver support.
Also, add a separate commit for non-functional changes (renaming and format changes)

dharunkrithikk committed:
drivers: spi: infineon: cleanup and rename

Rename Kconfig options and internal macros to remove "CAT1" specificity,
making the driver more generic for the Infineon family. This is a
non-functional change to prepare for PSoC 4 support.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk committed:
drivers: spi: infineon: add PSoC 4 support

Add support for the Infineon PSoC 4 family to the SPI driver,
Kconfig options, and polling-based transfer completion for slave
mode as PSoC 4 slave hardware does not control transaction start.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk committed:
drivers: spi: infineon: implement SCB block number lookup

Implement ifx_get_hw_block_num to retrieve the hardware block number
for a given SCB base address. This replaces the dependency on the
UART-specific lookup function ifx_cat1_uart_get_hw_block_num, making
the SPI driver more self-contained and ensuring correct resource
identification for SCB blocks.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

dharunkrithikk commented on drivers/spi/spi_infineon_pdl.c:
```diff
@@ -18,39 +18,37 @@ LOG_MODULE_REGISTER(cat1_spi, CONFIG_SPI_LOG_LEVEL);
 #include <cy_scb_spi.h>
 #include <cy_trigmux.h>
 
-#define IFX_CAT1_SPI_DEFAULT_OVERSAMPLE (4)
```
@sreeram yes, this is accessed through Kconfig, if any other board wants to change the oversample value based on user needs

sreeramIfx replied:
please move this to the appropriate commit as well

teburd commented on drivers/spi/spi_infineon_pdl.c:
```diff
@@ -142,7 +140,120 @@ static void ifx_cat1_spi_cb_wrapper(const struct device *dev, uint32_t event);
+{
+	uint32_t val;
+
+	for (val = 0u; val < IFX_SCB_ARRAY_SIZE; val++) {
```
Why didn't we stash the peripheral instance number in DT? Seems like we are doing at runtime what we could be doing at compile time here.

teburd reviewed (changes requested):
I'd say lets drop the support for peripheral (slave) mode in this PR to move forward as many of the issues that would be blockers here are around slave mode. We can revisit this mode in a future change.

dharunkrithikk commented on drivers/spi/spi_infineon_pdl.c:
```diff
@@ -1013,6 +1176,15 @@ static cy_rslt_t spi_init_hw(const struct device *dev, cy_stc_scb_spi_config_t *
+		 * (matching bare-metal initialization order)
+		 */
+		if (cfg->spiMode == CY_SCB_SPI_SLAVE) {
+			Cy_SCB_SPI_SetActiveSlaveSelect(config->reg_addr, CY_SCB_SPI_SLAVE_SELECT0);
```
Agreed,this should not be hard-wired and should depend on the DT configuration.
For now it matches the current board setup, this should be cleaned up and made configurable.

teburd reviewed (approved):
Looks much better to me. I really don't love this instance mapping thing we have going on, but I see we need to fix some other things to correct that.

dharunkrithikk commented on boards/infineon/cy8cproto_041tp/cy8cproto_041tp_common.dtsi:
```diff
@@ -23,6 +23,16 @@ uart0: &scb0 {
 	pinctrl-names = "default";
 };
 
+spi1: &scb1 {
```
Those are hardware cs pins that can be muxed directly, but I'm following the standard approach using 
`cs-gpios` for software-controlled CS. Thought it might be more flexible with how most zephyr SPI devices are configured.
